mat64/triangular.go

```diff
@@ -395,6 +396,76 @@ func (t *TriDense) MulTri(a, b Triangular) {
 	}
 }
 
+// Exp calculates the exponential of the matrix a, e^a, placing the result
+// in the receiver. If a has an odd number of rows, the matrix will be converted
+// to a Dense to perform the calculation.
+func (t *TriDense) Exp(a Triangular) {
+	n, kind := a.Triangle()
+	t.reuseAs(a.Triangle())
+
+	aU, _ := untransposeTri(a)
+	t.reuseAs(n, kind)
+	var restore func()
+	if t == aU {
+		t, restore = t.isolatedWorkspace(aU)
+		defer restore()
+	}
+	t.Copy(a)
+
+	if n&(n-1) == 0 {
+		// Exponential is trivial for sizes of powers of two.
+		// Calculation performed on upper triangular.
+		computeKind := matrix.Upper
+		aP := NewTriDense(n, computeKind, nil)
+		if kind != computeKind {
+			aP.Copy(a.T())
+		} else {
+			aP.Copy(a)
+		}
+		tUp := *NewTriDense(n, computeKind, nil)
```

Comment:
It seems like there are a lot of optimizations here. It'll require special casing for the computeKind, but I think even with the duplication we'll still be reducing lines of code. 

Big ideas:
1) Eye is just 1 or 0, so that's easily special cased in setting.
2) diagExp is just a diagonal, so it doesn't need a full TriDense worth of storage.
3) the multiplication with a diagonal is easy, it's just scaling the row.

I think it would look something like
````
if n%(n-1) == 0 {
    if t.kind == matrix.Upper {
         for i := 0; i < n; i++ {
              v := t.At(i,i)
              exp := math.Exp(v)
              t.Set(i,i, (1+v) * exp)
              for j := i+1; j < n; j++ {   // note the indexing optimization
                     v := t.At(i,j)                 
                     t.SetTri(i, j, v * exp)
              }
        }
}
````

With a similar code for the Lower case.